source/common/router/router.cc

```diff
@@ -1840,15 +1841,59 @@ bool Filter::convertRequestHeadersForInternalRedirect(Http::RequestHeaderMap& do
   const std::string original_host(downstream_headers.getHostValue());
   const std::string original_path(downstream_headers.getPathValue());
   const bool scheme_is_set = (downstream_headers.Scheme() != nullptr);
-  Cleanup restore_original_headers(
-      [&downstream_headers, original_host, original_path, scheme_is_set, scheme_is_http]() {
-        downstream_headers.setHost(original_host);
-        downstream_headers.setPath(original_path);
-        if (scheme_is_set) {
-          downstream_headers.setScheme(scheme_is_http ? Http::Headers::get().SchemeValues.Http
-                                                      : Http::Headers::get().SchemeValues.Https);
+
+  auto saved_headers = Http::RequestHeaderMapImpl::create();
```

Comment:
Honestly at this point if we need to restore original headers I think it'd be easier to just clone and restore, than try to unwind.  WDYT?

Reply:
If we stick with this let's make sure is thoroughly tested

Reply:
Yeah, that is less code, and easier to follow, I think. Done.